sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
-              jobIdToken, inputTable, extractDestinationDir, bqServices,
-              StaticValueProvider.of(executingProject));
+          // Create a singleton job ID token at execution time.
+          jobIdTokenCollection = p
+            .apply("TriggerIdCreation", Create.of("ignored"))
+            .apply("CreateJobId", MapElements.via(
+                new SimpleFunction<String, String>() {
+                  @Override
+                  public String apply(String input) {
+                    return randomUUIDString();
+                  }
+                }));
+          jobIdTokenView = jobIdTokenCollection
+              .apply(View.<String>asSingleton());
         }
+
+        PCollection<TableRow> rows;
+        if (true) {
+          // Apply the traditional Source model.
+          rows = p.apply(
+              org.apache.beam.sdk.io.Read.from(applySourceTransform(
+                  p, jobUuid, extractDestinationDir)))
+              .setCoder(getDefaultOutputCoder());
+        } else {
+          // Apply the DoFn version.
+          final TupleTag<String> filesTag =
+              new TupleTag<String>(){};
+          final TupleTag<TableSchema> tableSchemaTag =
+              new TupleTag<TableSchema>(){};
+          PCollectionTuple tuple = jobIdTokenCollection.apply("RunCreateJob", ParDo
+              .withOutputTags(filesTag, TupleTagList.of(tableSchemaTag)).of(
+                  new DoFn<String, String>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext c)
+                        throws Exception {
+                      String jobUuid = c.element();
+                      BigQuerySourceBase source = applySourceTransform(
+                          p, jobUuid, extractDestinationDir);
+                      TableSchema schema = source.getSchema(p.getOptions());
+                      c.sideOutput(tableSchemaTag, schema);
+                      List<String> files = source.extractFiles(p.getOptions());
+                      for (String file : files) {
+                        c.output(file);
+                      }
+                    }
+                  }));
+          final PCollectionView<TableSchema> schemaView = tuple.get(tableSchemaTag)
+              .apply(View.<TableSchema>asSingleton());
+          rows = tuple.get(filesTag)
+              //.apply(Reshuffle.<String, String>of())
+              .apply("ReadFiles", ParDo.of(
+                  new DoFn<String, TableRow>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext c)
+                        throws Exception {
+                      TableSchema schema = c.sideInput(schemaView);
+                      BigQuerySourceBase source = applySourceTransform(p, null, null);
+                      List<BoundedSource<TableRow>> sources =
+                          source.createSources(ImmutableList.of(c.element()), schema);
```

Comment:
I believe what is being suggested is that we can just output all of the created sources, and have a `DoFn` that reads from the `BoundedSource` we produce - and that way the runner can determine how to distribute that work.